### PR TITLE
feat: Image support getPopupContainer

### PR DESCRIPTION
**Pull request by rfreling**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [x] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

Add getPreviewContainer to Image Preview

rc-image PR react-component/image#29

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Add `getContainer` to Image preview property  |
| 🇨🇳 Chinese |  Image 支持 `preview.getContainer` 属性用于指定预览对话框的容器。    |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by rfreling** [comment]
Note: dependent on https://github.com/react-component/image/pull/29

**Comment by rfreling** [comment]
👋  @shaodahong 

After this suggestion: https://github.com/react-component/image/pull/29#discussion_r490757613

The build fails because of this type conflict: https://github.com/ant-design/ant-design/blob/5e0352200b6c72ffaaf7176ad64606d4e4faa147/components/config-provider/index.tsx#L31

--------

Which do you suggest?

(1) [ant-design] https://share.getcloudapp.com/z8uZN4Nr

or

(2) [react-component/Image] https://share.getcloudapp.com/Blu5rYgv

**Comment by shaodahong** [comment]
> 👋 @shaodahong
> 
> After this suggestion: [react-component/image#29 (comment)](https://github.com/react-component/image/pull/29#discussion_r490757613)
> 
> The [build fails](https://github.com/ant-design/ant-design/pull/26713/checks?check_run_id=1150660718) because of this type conflict:
> 
> https://github.com/ant-design/ant-design/blob/5e0352200b6c72ffaaf7176ad64606d4e4faa147/components/config-provider/index.tsx#L31
> 
> Which do you suggest?
> 
> (1) [ant-design] https://share.getcloudapp.com/z8uZN4Nr
> 
> or
> 
> (2) [react-component/Image] https://share.getcloudapp.com/Blu5rYgv

https://github.com/react-component/util/blob/69046a6bb2a111cb1909bf5b6a661b27a846e0f5/src/PortalWrapper.js#L30

I don't find `triggerNode`, @zombieJ `getPopupContainer?: (triggerNode: HTMLElement) => HTMLElement;` it's same `getContainer `?

File: components/image/index.tsx [hunk 1]
```diff
@@ -2,11 +2,15 @@ import * as React from 'react';
...
-const Image: React.FC<ImageProps> = ({ prefixCls: customizePrefixCls, ...otherProps }) => {
-  const { getPrefixCls }: ConfigConsumerProps = React.useContext(ConfigContext);
+const Image: React.FC<ImageProps> = ({ prefixCls: customizePrefixCls, getPopupContainer, ...otherProps }) => {
+  const { getPrefixCls, getPopupContainer: getContextPopupContainer }: ConfigConsumerProps = React.useContext(ConfigContext);
   const prefixCls = getPrefixCls('image', customizePrefixCls);
 
-  return <RcImage prefixCls={prefixCls} {...otherProps} />;
+  return <RcImage 
+           prefixCls={prefixCls} 
+           getPopupContainer={getPopupContainer === undefined ? getContextPopupContainer : getPopupContainer}
```
Comment (zombieJ) [inline components/image/index.tsx:11]: `getPopupContainer === undefined ? getContextPopupContainer : getPopupContainer` -> `getPopupContainer ?? getContextPopupContainer`

**Comment by zombieJ** [comment]
Also need update doc.

File: components/image/index.en-US.md [hunk 2]
```diff
@@ -24,5 +24,6 @@ Previewable image.
 | preview | Whether to enable the preview | boolean | true | 4.6.0 |
 | src | Image path | string | - | 4.6.0 |
 | width | Image width | string \| number | - | 4.6.0 |
+| getPopupContainer | The DOM container of the preview dialog | function | () => document.body | 4.7.0 |
```
Comment (zombieJ) [inline components/image/index.en-US.md:27]: `function` => `() => HTMLElement`

**Review by zombieJ — APPROVED** [review]

**Comment by rfreling** [comment]
> /rebase

@zombieJ Thanks for reviews. The build is failing because of the type conflict here: https://github.com/ant-design/ant-design/pull/26713#issuecomment-697078743

Do you have a suggestion for how to handle before rebasing?

**Comment by zombieJ** [comment]
You can ref https://ant.design/components/modal-cn/#header. Preview use `rc-dialog` internal. It may no need pass triggerNode currently.

File: components/image/index.tsx [hunk 3]
```diff
@@ -2,11 +2,15 @@ import * as React from 'react';
 import RcImage, { ImageProps } from 'rc-image';
 import { ConfigConsumerProps, ConfigContext } from '../config-provider';
 
-const Image: React.FC<ImageProps> = ({ prefixCls: customizePrefixCls, ...otherProps }) => {
-  const { getPrefixCls }: ConfigConsumerProps = React.useContext(ConfigContext);
+const Image: React.FC<ImageProps> = ({ prefixCls: customizePrefixCls, getPopupContainer, ...otherProps }) => {
+  const { getPrefixCls, getPopupContainer: getContextPopupContainer }: ConfigConsumerProps = React.useContext(ConfigContext);
```
Comment (rfreling) [inline components/image/index.tsx:6]:
```suggestion
  const { getPrefixCls }: ConfigConsumerProps = React.useContext(ConfigContext);
```

File: components/image/index.tsx [hunk 4]
```diff
@@ -2,11 +2,15 @@ import * as React from 'react';
...
   const prefixCls = getPrefixCls('image', customizePrefixCls);
 
-  return <RcImage prefixCls={prefixCls} {...otherProps} />;
+  return <RcImage 
+           prefixCls={prefixCls} 
+           getPopupContainer={getPopupContainer ?? getContextPopupContainer}
```
Comment (rfreling) [inline components/image/index.tsx:11]:
```suggestion
           getPopupContainer={getPopupContainer}
```

**Review by rfreling — COMMENTED** [review]
@zombieJ These suggested changes will fix the type conflict so that the build will pass.

I was following the patterns here which somehow passes the type check:

https://github.com/ant-design/ant-design/blob/006abe5e92007e6c01d5137bc8ba6750cb2b0302/components/modal/Modal.tsx#L130

https://github.com/ant-design/ant-design/blob/006abe5e92007e6c01d5137bc8ba6750cb2b0302/components/modal/Modal.tsx#L198

**Comment by zombieJ** [comment]
CI failed

**Comment by afc163** [comment]
这个属性最好也移到 `preview: { getContainer }` 里面去。

File: components/image/index.en-US.md [hunk 5]
```diff
@@ -24,5 +24,6 @@ Previewable image.
 | preview | Whether to enable the preview | boolean | true | 4.6.0 |
 | src | Image path | string | - | 4.6.0 |
 | width | Image width | string \| number | - | 4.6.0 |
+| getPopupContainer | The DOM container of the preview dialog | () => HTMLElement | () => document.body | 4.7.0 |
```
Comment (hengkx) [inline components/image/index.en-US.md:27]: 4.8.0

**Comment by afc163** [comment]
wait for https://github.com/react-component/image/pull/34

**Comment by hengkx** [comment]
https://github.com/react-component/image/pull/34#issuecomment-708304230

File: components/image/index.zh-CN.md [hunk 6]
```diff
@@ -25,14 +25,14 @@ cover: https://gw.alipayobjects.com/zos/antfincdn/D1dXz9PZqa/image.svg
...
 | width | 图像宽度 | string \| number | - | 4.6.0 |
-| getPopupContainer | 指定 Preview 挂载的 HTML 节点 | () => HTMLElement | () => document.body | 4.8.0 |
 
 ### previewType
 
 ```
 {
   visible: boolean,
-  onVisibleChange:function(value, prevValue)
+  onVisibleChange:function(value, prevValue),
```
Comment (afc163) [inline components/image/index.zh-CN.md]:
```suggestion
  onVisibleChange: (value, prevValue) => void,
```

File: components/image/index.en-US.md [hunk 7]
```diff
@@ -24,14 +24,14 @@ Previewable image.
...
-| getPopupContainer | The DOM container of the preview dialog | () => HTMLElement | () => document.body | 4.8.0 |
 
 ### previewType
 
 ```
 {
   visible: boolean,
-  onVisibleChange:function(value, prevValue)
+  onVisibleChange:function(value, prevValue),
+  getContainer: string | HTMLElement | (() => HTMLElement)
```
Comment (afc163) [inline components/image/index.en-US.md]: string 和 HTMLElement 支持么？

**Comment by afc163** [comment]
You have to upgrade rc-image to 4.0.0

**Comment by hengkx** [comment]
Please check 
image

**Comment by afc163** [comment]
Please fill changelog part in PR.

**Comment by rfreling** [comment]
> Please fill changelog part in PR.

The Chinese is my best guess – will gladly accept edits :)

image

**Comment by afc163** [comment]
> The Chinese is my best guess – will gladly accept edits :)

You could leave chinese changelog to us. 😹

**Pull request merged**